File: jacobi.cpp
#include <fstream>
#include <iostream>
#include <sstream>
#include <string>
#include <vector>
#include <cmath>

using namespace std;

vector<double> xValues,xNewValues;
double**matrix;

int initializeMatrix(string origin){
    ifstream read;
    read.open(origin.c_str());
    int size = 0;
    read >> size;
    matrix = new double *[size];
    for (int i = 0; i < size; ++i){
        matrix[i] = new double [size+1];
    }

    for (int i = 0; i < size; ++i){
        for (int j = 0; j < size + 1; ++j){
            read >> matrix[i][j];
        }
    }
    return size;
}


double newJacobi(int matrixSize){
    double suma, disp,var, aii;
    disp = 0;
    for (int i = 0; i < matrixSize;++i){
        suma = 0;
        for (int j = 0; j < matrixSize;++j){
            var = matrix[i][j];
            if (i!=j)
                suma += var*xValues[j];
            else
                aii = var;
        }
        var = matrix[i][matrixSize];
        xNewValues[i] = (var - suma)/aii;
        disp = max(disp, abs(xNewValues[i]- xValues[i]));
    }
    return disp;
}

bool jacobi(long long matrixSize, double tol, long long niter){
    double disp = tol+1;
    int cont = 0;

    while (disp > tol && cont < niter){
        for (int i = 0; i < matrixSize; ++i)xValues[i] = xNewValues[i];
        disp = newJacobi(matrixSize);
        cont++;
    }

    if (disp <= tol)
        return true;
    return false;
}


void writeSolution(string outputFile){
    ofstream write;
    write.open(outputFile.c_str(),ios::trunc);
    for (int i = 0; i < xValues.size();++i){
        write << xValues[i] << endl;
    }
    write.close();
}

int main(){
    std::ios::sync_with_stdio(false);
    string matrixInputFile, outputFile;
    long long matrixSize, maxIterations;
    double tolerance;
    bool success;

    matrixInputFile = "matrix.txt";
    outputFile = "sol.txt";

    matrixSize = initializeMatrix(matrixInputFile);

    xValues.assign(matrixSize, 0);
    xNewValues.assign(matrixSize, 0);

    cout << "how much tolerance? for example if you have n=10, tolerance should be 0.01 and maxIters=10000" << endl;
    cin >> tolerance;
    cout << "how many iterations? i.e: 100" << endl;
    cin >> maxIterations;

    success = jacobi(matrixSize, tolerance, maxIterations);
    
    if (success)
        writeSolution(outputFile);
    else
        cout << "could not reach the solutions in " << maxIterations << " iterations" << endl;
        
    return 0;
}

File: InterfaceMethodsPython/jacobi.py
import numpy as np
import sys
import json
from math import *



def initializeMatrix(file,size):
    A = np.zeros((size, size+1))
     #sys.argv[4] array con los valores iniciales
    xValues= json.loads(sys.argv[4])

    xNewValues = np.zeros(size)
    for i in range (size):
        cont=0
        fileMio=file.readline().replace("\n","").split(" ")
        linea=fileMio
        for j in linea:
            A[i][cont]=float(j)
            cont+=1
    return A,xValues,xNewValues

def newJacobi(xValues,xNewValues,A,size):
    disp = 0
    for i in range (size):
        suma = 0
        for j in range (size):
            var = A[i][j]
            if (i!=j):
                suma += var*xValues[j]
            else:
                aii = var

        var = A[i][size]
        xNewValues[i] = (var - suma)/aii
        disp = max(disp, abs(xNewValues[i]- xValues[i]))

    return disp,xNewValues

def jacobi(xValues,xNewValues,A,size,tol,niter):
    disp = tol+1
    cont = 0
    while (disp > tol and cont < niter):
        for i in range (size):
            xValues[i] = xNewValues[i]
        disp,xNewValues = newJacobi(xValues,xNewValues,A,size)
        print "Iter:",cont
        cont+=1

        print "X vector:",np.array(xNewValues)
        print ("\n")
        print "Error: ",disp
        print ("\n")

    if (disp <= tol):
        bandera= True
    else:
        bandera=False
    return bandera,xValues,xNewValues,A,tol,niter,disp

def main():
    name=sys.argv[1]
    file=open(name)
    size=int(file.readline())
    A,xValues,xNewValues = initializeMatrix(file,size)

    #tolerance=float(input( "how much tolerance? for example if you have n=10, tolerance should be 0.01 and maxIters=10000\n"  ))
    tolerance=float(sys.argv[2])
    #maxIterations=float(input("how many iterations? i.e: 100\n"))
    maxIterations=float(sys.argv[3])
    success,xValues,xNewValues,A,tol,niter,error = jacobi(xValues,xNewValues,A,size, tolerance, maxIterations)
    #print success,xValues,xNewValues,A,tol,niter
    if (success):
        print ("Matrix A")
        print (np.array(A))
        print ("\n")
        print np.array(xValues)
        print ("\n")
        print "Error: ",error
        print ("\n")

    else:
        print "could not reach the solutions in ", maxIterations , " iterations"

main()
